src/hotspot/share/opto/superword.cpp

```diff
@@ -1946,16 +1966,16 @@ Node_List* CMoveKit::make_cmovevd_pack(Node_List* cmovd_pk) {
       || cmpd->in(0) != NULL  // CmpDNode has control flow!!
       || _sw->my_pack(cmpd) == NULL) {
       NOT_PRODUCT(if(_sw->is_trace_cmov()) {tty->print("CMoveKit::make_cmovevd_pack: CmpD %d does not fit CMoveD %d for building vector, escaping...", cmpd->_idx, cmovd->_idx); cmpd->dump();})
-      return NULL;
+    return cmovd_pk;
   }
   Node_List* cmpd_pk = _sw->my_pack(cmpd);
   if (cmpd_pk->size() != cmovd_pk->size() ) {
-    return NULL;
+    return cmovd_pk;
   }
 
   if (!test_cmpd_pack(cmpd_pk, cmovd_pk)) {
     NOT_PRODUCT(if(_sw->is_trace_cmov()) {tty->print("CMoveKit::make_cmovevd_pack: cmpd pack for CmpD %d failed vectorization test", cmpd->_idx); cmpd->dump();})
-    return NULL;
+    return cmovd_pk;
   }
 
   Node_List* new_cmpd_pk = new Node_List();
```

Comment:
The following suggestion is just an idea as I was a little bit confused by how you use the return value of `make_cmovevd_pack` to remove the cmove pack and its related packs. Intuitively, I would have expected that this "make method" returns the newly created pack instead. 

Maybe it's cleaner if you split this method into a "should merge" method with the check
```
 if ((cmovd->Opcode() != Op_CMoveF && cmovd->Opcode() != Op_CMoveD) ||
      pack(cmovd) != NULL /* already in the cmov pack */) {
    return NULL;
  }
```
a "can merge" method that checks all the other constraints and an actual "make pack" method with the code starting at this line. Then you could use these methods in `merge_packs_to_cmovd` like that in pseudo-code:
```
void SuperWord::merge_packs_to_cmovd() {
  for (int i = _packset.length() - 1; i >= 0; i--) {
    Node_List* pack = _packset.at(i);
    if (_cmovev_kit.should_merge(pack)) {
      if (_cmovev_kit.can_merge(pack)) {
        _cmovev_kit.make_cmovevd_pack(pack)
      } else {
        remove_cmove_and_related_packs(pack);
      }
    }
  }
  ...
```

Reply:
@chhagedorn thanks for your great suggestion. It did make the code much easier to understand. I've done the refactoring in the latest commit. Please help review. Thanks for your time!